library/psa_its_file.c

```diff
@@ -214,9 +214,12 @@ psa_status_t psa_its_set( psa_storage_uid_t uid,
     n = fwrite( &header, 1, sizeof( header ), stream );
     if( n != sizeof( header ) )
         goto exit;
-    n = fwrite( p_data, 1, data_length, stream );
-    if( n != data_length )
-        goto exit;
+    if( p_data != NULL )
```

Comment:
Is this enough? The Doxygen for this function says, among others, it will return `PSA_ERROR_INVALID_ARGUMENT`, if:

> The operation failed because one of the provided pointers(`p_data`) is invalid, for example is `NULL` or references memory the caller cannot access.

If the docs are up to date, I suggest returning the above error.

Reply:
It is not required for the implementation to report this error code in the case of an invalid memory reference, as a memory access violation is also a reasonable outcome of such a call. But an implementation is permitted to detect and report invalid memory references using this error code.

It would be helpful if the spec provided guidance on how memory buffers are handled in the API, for comparison see the PSA-FF and PSA Crypto specifications.

In this API `p_data` should be permitted to have any value if `data_length==0`. In that specific case:
* `fwrite()` has fully defined behaviour (do nothing)
* `psa_its_set()` has agreed behaviour (but it is not explicit in the spec), when it will create a zero-length data asset.

However, I can see that avoiding the potential UB when `(data_length > 0 && p_data == NULL)` is important due to the risk of leaving a half-populated ITS storage file in the system. So maybe you need:

```
if (p_data == NULL && data_length > 0)
    return PSA_ERROR_INVALID_ARGUMENT;
```

further up the function, before creating/opening the file and storing any data.

Reply:
I assumed that creating a file with a good header but no data is a valid case. Adding such check will change behaviour - is this what we want?

Reply:
`data_length == 0 && p_data == NULL` is a valid use case. It creates an empty file. Crypto is unlikely to need it, but this implementation of ITS is meant to be generic.

`data_length != 0 && p_data == NULL` is not a valid input. In the PSA Crypto API, there is implementation guidance to use the normal handling of invalid pointer dereferences (typically terminating the application and if possible logging the fault), and not return an error from the function. (The rationale for this guidance is that returning an error has no benefit since it's a programmer error and not a recoverable error, it makes errors harder to debug, it confuses static analyzers that have heuristics around nullable pointers by making them think that a null input is ok, and the code to check for null is pure overhead with no benefit.) I recommend following this guidance elsewhere in PSA.